cupyx/scipy/sparse/compressed.py

```diff
@@ -544,30 +544,22 @@ def _major_index_fancy(self, idx):
     def _minor_index_fancy(self, idx):
         """Index along the minor axis where idx is an array of ints.
         """
-
-        idx_dtype = self.indices.dtype
-        idx = cupy.asarray(idx, dtype=idx_dtype).ravel()
-
         M, N = self._swap(*self.shape)
-        k = len(idx)
+        k = idx.size
         new_shape = self._swap(M, k)
         if k == 0:
             return self.__class__(new_shape)
 
-        # pass 1: count idx entries and compute new indptr
-        col_order = cupy.argsort(idx).astype(idx_dtype, copy=False)
-
-        index1_outs = _index._csr_column_index1(idx, self.indptr, self.indices)
-        res_indptr, indices_mask, col_counts, sort_idxs = index1_outs
+        if idx.size * M < self.nnz:
+            # TODO (asi1024): Implement faster algorithm.
+            pass
 
-        # pass 2: copy indices/data for selected idxs
-
-        res_indices, res_data = _index._csr_column_index2(
-            col_order, col_counts, sort_idxs, self.indptr, indices_mask,
-            self.data, res_indptr)
-
-        return self.__class__((res_data, res_indices, res_indptr),
-                              shape=new_shape, copy=False)
+        name = type(self).__name__
+        if name == 'csr_matrix':
+            return self.tocsc()._major_index_fancy(idx).tocsr()
```

Comment:
My only concern here would be memory ussage, but since these are sparse matrices probably their memory footprint is almost negligible in the 99% of the use cases.

Reply:
The trade-off between memory and speed is what prompted the original design.

We have folks that are using tall and extremely wide datasets for NLP and selecting a few / subset of columns would require converting to a csc with an index on the order of the number of columns.

Reply:
Thanks! that makes complete sense!!

Maybe we can take a hybrid approach based on the shape or size of the matrix?

Reply:
I confirmed that this change reduces memory usage even when indexing with few columns.

- master
```
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 7423 14753], format = csr, density = 0.01, dtype = float32, n_rows = 15000, n_cols = 15000
67569664
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 7423 14753], format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
3375244288
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 3297 10656  3876 ... 11454  4865  3201], format = csr, density = 0.01, dtype = float32, n_rows = 15000, n_cols = 15000
74048000
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 3297 10656  3876 ... 11454  4865  3201], format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
3693047808
```

- This PR
```
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [1620 3058], format = csr, density = 0.01, dtype = float32, n_rows = 15000, n_cols = 15000
45314560
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [1620 3058], format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
2250314240
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 8267  2341  7173 ...   191  9984 14152], format = csr, density = 0.01, dtype = float32, n_rows = 15000, n_cols = 15000
45330432
- name = minor_column_fancy_indexing, row = slice(None, None, None), col = [ 8267  2341  7173 ...   191  9984 14152], format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
2250330112
```

 <details>
- Reproducer

```py
import cupy
from cupy import random
import cupyx.time


class MyMemoryHook(cupy.cuda.MemoryHook):
    def __init__(self):
        self.allocated_size = 0
        self.max_allocated_size = 0

    def malloc_preprocess(self, device_id, size, mem_size):
        self.allocated_size += mem_size
        self.max_allocated_size = max(self.max_allocated_size, self.allocated_size)

    def free_preprocess(self, device_id, mem_size, mem_ptr, pmem_id):
        self.allocated_size -= mem_size


def index(array, row, col):
    if min is None:
        return array[row]
    else:
        return array[row, col]


def run(name, row, col, format, density, dtype, n_rows, n_cols):
    a = cupy.sparse.random(n_rows, n_cols, format=format, density=density)
    a = a.astype(dtype)
    print('- name = {}, row = {}, col = {}, format = {}, density = {}, '
          'dtype = {}, n_rows = {}, n_cols = {}'.format(
              name, row, col, format, density, dtype, n_rows, n_cols))
    with MyMemoryHook() as hook:
        index(a, row, col)
        print(hook.max_allocated_size)


if __name__ == '__main__':
    params = [
        ('minor_column_fancy_indexing', 'csr', slice(None), random.randint(0, 14999, size=2)),
        ('minor_column_fancy_indexing', 'csr', slice(None), random.randint(0, 14999, size=4000)),
    ]
    for name, format, row, col in params:
        for density in [0.01, 0.5]:
            for dtype in ['float32']:
                for n_rows in [15000]:
                    for n_cols in [15000]:
                        run(name, row, col, format, density, dtype, n_rows, n_cols)
```

Reply:
Anyway, I think that there is a better algorithm for small indices. (commented in https://github.com/cupy/cupy/pull/3886/files#diff-91edca6879a46de14928e1d9107fd8adR554)

Reply:
Ah, that's right. Now I recall the scipy implementation was always using an inverted index of size n_cols. 

I was able to shrink that down, but only by a little bit. If this is both faster and uses less memory then that's a no-brainer. It also removes quite a bit of complexity.